File: afbilling/src/main/java/com/appyfurious/validation/ValidationCallback.java
package com.appyfurious.validation;

import android.content.Context;
import android.support.annotation.NonNull;

import com.appyfurious.validation.event.FacebookInteractor;

import org.jetbrains.annotations.NotNull;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * ValidationCallback.java
 * getfitandroid
 * <p>
 * Created by o.davidovich on 25.05.2018.
 * <p>
 * Copyright © 2018 Appyfurious. All rights reserved.
 */

public class ValidationCallback implements Callback<ResponseBody> {

    private CryptoAES128 mEncryptor;
    private ValidationListener mValidationListener;

    private String mSecretKey;

    public ValidationCallback(@NotNull String secretKey, ValidationListener listener) {
        this.mSecretKey = secretKey;
        this.mValidationListener = listener;
    }

    @Override
    public void onResponse(@NonNull Call<ResponseBody> call, @NonNull Response<ResponseBody> response) {
        if (response.isSuccessful()) {

            mEncryptor = new CryptoAES128(mSecretKey);
            String decryptString;
            try {
                decryptString = mEncryptor.decrypt(response.body().string());
                JSONObject jsonResponse = new JSONObject(decryptString);
                JSONObject jsonData = jsonResponse.getJSONObject("data");
                boolean isValid = jsonData.getBoolean("isValid");
                if (isValid) {
                    validationSuccess();
                } else {
                    mValidationListener.onValidationFailure("Validation Error");
                }
            } catch (IOException e) {
                validationSuccess();
            } catch (JSONException e) {
                validationSuccess();
            }
        } else {
            validationSuccess();
        }
    }

    @Override
    public void onFailure(@NonNull Call<ResponseBody> call, @NonNull Throwable t) {
        validationSuccess();
    }

    private void validationSuccess() {
        mValidationListener.onValidationSuccess();
        FacebookInteractor.logAddedToCartEvent(mValidationListener.getValidationContext());
    }

    public interface ValidationListener {

        Context getValidationContext();

        void onValidationSuccess();

        void onValidationFailure(String errorMessage);
    }

}

File: afbilling/src/main/java/com/appyfurious/validation/ValidationCallback.kt
package com.appyfurious.validation

import com.appyfurious.log.Logger
import okhttp3.ResponseBody
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class ValidationCallback(private val secretKey: String,
                         private val validationListener: ValidationListener? = null,
                         private val restoreListener: RestoreListener? = null)
    : Callback<ResponseBody> {

    private var encryptor: CryptoAES128? = null

    override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
        val body = response.body()?.string()
        when {
            response.code() in 500..599 -> {
                Logger.notify("response.code() in 500..599 validationSuccess")
                validationSuccess()
            }
            else -> {
                encryptor = CryptoAES128(secretKey)
                try {
                    val decryptString = encryptor?.decrypt(body)
                    Logger.notify("isNotNull: ${decryptString != null}, decryptString: $decryptString")
                    val jsonResponse = JSONObject(decryptString)
                    Logger.notify(jsonResponse.toString())
                    val jsonData = jsonResponse.getJSONObject("data")
                    Logger.notify(jsonData.toString())
                    val isValid = jsonData.getBoolean("isValid")
                    Logger.notify(isValid.toString())
                    if (isValid) {
                        Logger.notify("isValid validationSuccess")
                        validationSuccess()
                    } else {
                        Logger.notify("isValid == false onValidationFailure")
                        validationFailure()
                    }
                } catch (e: Exception) {
                    Logger.exception("Exception onValidationFailure")
                    Logger.exception(e)
                    validationFailure("Exception onValidationFailure")
                }
            }
        }
        validationListener?.onValidationHideProgress()
        Logger.notify("response body: $body")
        Logger.notify("response all: $response")
    }

    override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
        Logger.notify("onFailure validationFailure")
        validationFailure("onFailure validationFailure")
        validationListener?.onValidationHideProgress()
    }

    private fun validationSuccess() {
        validationListener?.onValidationSuccess()
        restoreListener?.validationRestoreSuccess()
    }

    private fun validationFailure(errorMessage: String = "Validation Error") {
        validationListener?.onValidationFailure(errorMessage)
        restoreListener?.validationRestoreFailure(errorMessage)
    }

    interface ValidationListener {
        fun onValidationSuccess()
        fun onValidationFailure(errorMessage: String)
        fun onValidationShowProgress()
        fun onValidationHideProgress()
    }

    interface RestoreListener {
        fun validationRestoreSuccess()
        fun validationRestoreFailure(errorMessage: String)
    }
}